### DISCUSSION TITLE
How to port MicroPython to a brand new processor?

### DISCUSSION THREAD

**Discussion by ijoffe** [micropython/micropython#13438] [General]
I am part of a group looking to get MicroPython running on a new proprietary board with a custom processor (no existing port).

I have seen the [documentation for porting](https://github.com/micropython/micropython/blob/master/docs/develop/porting.rst), but are these instructions for porting to a brand new processor or just to a new board?

Is there a guide somewhere on porting MicroPython to a brand new processor (with a different, proprietary ISA)? What steps are required? Is this generally recommended?

Thanks!

**Comment by jimmo** [comment]
The first main question is what's the compiler and general build system and SDK/HAL for your custom processor.

So for example, if you're required to use non-Make-based build system then you're going to be in for a much harder time. MicroPython works with a range of C compilers, but obviously gcc is going to be the easiest. We don't use any extensions, any C99 compiler should work.

If your system has some sort of RTOS or other similar framework, then you might want to look at the `embed` port (i.e. embed MicroPython into your existing firmware) rather than write a direct "bare metal" port. (See also https://github.com/micropython/micropython/pull/11430)

Most of the complexity of a new port comes from figuring out how to integrate with the vendor SDK/HAL and getting the makefile working.

You can look at the `bare-arm` and `minimal` ports, but also some of the newer ports (e.g. `mimxrt`). The "reference" port is still `stm32`.

One thing to note, the ISA shouldn't be important (unless you have some specific boot/reset/interrupt config), nothing in MicroPython requires assembly code, except for:
- the NLR (exception handling), but if your platform supports `longjmp` you can default to that
- the native emitter, but you don't need to enable that

It's hard to give more details without really understanding where you're at. Feel free to ask more questions!

**Reply by ijoffe** [reply]
Thanks for the quick and detailed response!

Our processor currently does not have any native C/C++ compiler, we use cross-compiling to program C/C++ executables on the board. Does MicroPython require a native C/C++ compiler, or is there a way to port it without this?

**Reply by jimmo** [reply]
@ijoffe This is completely expected -- almost every target supported by MicroPython is using a cross compiler. (The only exception is the Unix and Windows ports)

**Comment by NoosaHydro** [comment]
You install the toolchain for that CPU on your laptop, jump into the micropython "ports" folder and pick the existing one closest to the one you've got, "cp -a" that into a new tree for your architecture, jump in there and run "make", then work your way through fixing everything that's not working until it builds.

Since you didn't name the processor, and asked about a "native C/C++ compiler", it looks like you don't understand anywhere near enough about MCU development to pull this off yourself though.  The good news though, is that this lack of knowledge of yours probably means that there is already a port that will almost work for you - you're just not aware of the name for it (e.g. an "ESP32" is an Xtensa CPU)

If you want anyone to help you, you have to tell us what we need to know in order to do that, so - spill the beans: what is so-called "custom processor" ?

**Comment by jimmo** [comment]
@NoosaHydro your reply (now hidden) was highly inappropriate and a violation of our code of conduct. https://github.com/micropython/micropython/blob/master/CODEOFCONDUCT.md

Please reconsider your approach in further interactions with the MicroPython project.

**Reply by NoosaHydro** [reply]
I strongly disagree with your objection to my long and helpful reply, and I further do not accept that it's your place to act on behalf of someone else who almost certainly did not object, and probably took no offence either. 

porting to a new architecture is very complicated, and this user (a 4th year unit student - I checked) is unaware of what they don't know, so I am doing them a valuable service by telling them that.  I have 40 *years* porting experience under my belt (PDP-11 being my first).

You can't go around assuming that when you tell someone they don't know something, that it's an insult NOT IT IS NOT.  It's HELPING THEM.

Please un-hide my response.  It's useful not just to ijoffe.  If ijoffe is in any way offended, they can speak up and I can apologise - but I do not think they will be, and we can't help them if they don't answer my question.
